### DISCUSSION TITLE
interface vs type

### DISCUSSION THREAD

**Discussion by AjaxSolutions** [pmndrs/jotai#1765] [Q&A] [answered]
I noticed that Jotai mostly uses types and there are very few interfaces used.  I'm not trying to start any heavy duty discussion about the merits of one approach vs another approach, but I'm just curious why interfaces are less favored.

The official TypeScript docs now recommend interfaces over Types
https://github.com/microsoft/TypeScript/wiki/Performance#preferring-interfaces-over-intersections

**Comment by dai-shi** [comment, marked as answer]
type aliases are just aliases. it's not to define something, but just to be short.
interfaces are APIs.
what we found is, as a library, that interfaces should be exported. otherwise something goes wrong.
jotai API is designed to be minimal and it also avoids to export types (it's the design decision for API types). thus, for internal types, type aliases are preferred.

**Reply by MGray55** [reply]
Are there actual problems that can happen with interfaces?
Is it OK if existing components use interfaces with Jotai?
Or is it just preferred in Jotai?
Thanks.

**Reply by dai-shi** [reply]
The problem happens if an interface is not exported by a library, and if an app tries to create a type definition file that depends on the interface, IIRC.  In your app code, there's no such restriction. It's just about the library design.
